vscode/src/testExplorer.ts

```diff
@@ -0,0 +1,234 @@
+
+  ctrl.createRunProfile(
+    "Run Tests",
+    vscode.TestRunProfileKind.Run,
+    runHandler,
+    true,
+    undefined,
+    false,
+  );
+
+  ctrl.resolveHandler = async (item) => {
+    if (!item) {
+      context.subscriptions.push(...startWatchingWorkspace(ctrl, context));
+      return;
+    }
+  };
+
+  function updateNodeForDocument(e: vscode.TextDocument) {
+    if (!isQsharpDocument(e)) {
+      return;
+    }
+  }
+
+  for (const document of vscode.workspace.textDocuments) {
+    updateNodeForDocument(document);
+  }
+
+  context.subscriptions.push(
+    vscode.workspace.onDidOpenTextDocument(updateNodeForDocument),
+    vscode.workspace.onDidChangeTextDocument((e) =>
+      updateNodeForDocument(e.document),
+    ),
+  );
+}
+
+/**
+ * If there are no workspace folders, then we can't watch anything. In general, though, there is a workspace since this extension
+ * is only activated when a .qs file is opened.
+ **/
+
+function getWorkspaceTestPatterns() {
+  if (!vscode.workspace.workspaceFolders) {
+    return [];
+  }
+
+  return vscode.workspace.workspaceFolders.map((workspaceFolder) => ({
+    workspaceFolder,
+    pattern: new vscode.RelativePattern(workspaceFolder, "**/*.qs"),
+  }));
+}
+
+/**
+ * Watches *.qs files and triggers the test discovery function on update/creation/deletion, ensuring we detect new tests without
+ * the user having to manually refresh the test explorer.
+ **/
+function startWatchingWorkspace(
+  controller: vscode.TestController,
+  context: vscode.ExtensionContext,
+) {
+  return getWorkspaceTestPatterns().map(({ pattern }) => {
+    const refresher = mkRefreshHandler(controller, context);
+    const watcher = vscode.workspace.createFileSystemWatcher(pattern);
+    watcher.onDidCreate(async () => {
+      await refresher();
+    });
+
+    watcher.onDidChange(async () => {
+      await refresher();
+    });
+
+    watcher.onDidDelete(async () => {
+      await refresher();
+    });
+
+    return watcher;
+  });
+}
+
+/**
+ * Given a single test case, run it in the worker (which runs the interpreter) and report results back to the
+ * `TestController` as a side effect.
+ *
+ * This function manages its own event target for the results of the test run and uses the controller to render the output in the VS Code UI.
+ **/
+async function runTestCase(
+  ctrl: vscode.TestController,
+  testCase: vscode.TestItem,
+  request: vscode.TestRunRequest,
+  worker: ICompilerWorker,
+  program: ProgramConfig,
+): Promise<void> {
+  if (testCase.children.size > 0) {
+    for (const childTestCase of testCase.children) {
+      await runTestCase(ctrl, childTestCase[1], request, worker, program);
+    }
+    return;
+  }
+  const run = ctrl.createTestRun(request);
+  const evtTarget = new QscEventTarget(false);
+  evtTarget.addEventListener("Message", (msg) => {
+    run.appendOutput(`Test ${testCase.id}: ${msg.detail}\r\n`);
+  });
+
+  evtTarget.addEventListener("Result", (msg) => {
+    if (msg.detail.success) {
```

Comment:
1. I think listeners need to be removed with `removeEventListener` after the test case is run? (If you copied the debugger code, I think the same applies there too)

2. Just checking, do you want to handle `"DumpMachine"` and `"DumpMatrix"` too, or is it intentionally filtered to messages?

3. (BTW, there is a more performant `uiresultsrefresh` event, which is an alternative to handling these individual events, that batches message/dumpmachine/results outputs so you don't have a million callbacks into JS if you're, for example, writing `Message()` in a loop from Q#. If performance becomes a concern that's worth looking into. I don't necessarily think it has to be done right now though).

Reply:
I decided to remove these manual listeners and reuse the debugger event target, which has all of the dumpmachine rendering logic built-in. While doing this, I found a bug in that renderer for `Result` types and fixed that as well.